### DISCUSSION TITLE
Testing expectEmit() and facing issues when Event is being used from an interface

### DISCUSSION THREAD

**Discussion by devrajsinghrawat** [Cyfrin/foundry-full-course-cu#554] [Q&A] [answered]
Hi Friends,

I am trying to test the expectEmit in below test case.

```
    function testFundSucessWithEnoughEth() public {
    // Fund method SHOULD received funds value of:  1000000000000000000 (1 ETH) 
    // from: 0x7FA9385bE102ac3EAc297483Dd6233D62b3e1496 (.i.e. address(this)).
        vm.expectEmit();
        emit IFundMe.FUNDED(address(this), SEND_VALUE);
        fundMe.fund{value: SEND_VALUE}();
    }
```

and for that I need to use emit event so instead of declaring the same event at two places I had created an interface and trying to use the same interface at FundMe as well in FundMeTest contracts

![Screenshot 2023-08-06 at 12 41 43 PM](https://github.com/Cyfrin/foundry-full-course-f23/assets/44597990/d0b91cb7-736b-467d-9517-a45f4d724da8)

Interface looks like below 

![Screenshot 2023-08-06 at 12 46 31 PM](https://github.com/Cyfrin/foundry-full-course-f23/assets/44597990/fd96df42-4362-4989-ba91-54a77050ffac)

Interface is imported in both places  .i.e. at FundMe and FundMeTest

![Screenshot 2023-08-06 at 12 43 41 PM](https://github.com/Cyfrin/foundry-full-course-f23/assets/44597990/0524645a-e92a-4530-ad79-17c80eb18092)

![Screenshot 2023-08-06 at 12 46 01 PM](https://github.com/Cyfrin/foundry-full-course-f23/assets/44597990/5af1d3e4-6757-4d46-920b-f4f9e2359038)

When only event is included in the interface then it works perfectly 

![Screenshot 2023-08-06 at 12 48 01 PM](https://github.com/Cyfrin/foundry-full-course-f23/assets/44597990/52a183bb-b21a-49fa-ae1b-dc5fde95f5ff)

![Screenshot 2023-08-06 at 12 48 29 PM](https://github.com/Cyfrin/foundry-full-course-f23/assets/44597990/734ee27c-887b-45ad-a230-a7862d158838)

But as soon as I also include the function signature in the interface then the test contract asks me in make it abstract contract caz it is not implementing the methods inside the interface and after I make it abstract then nothing works

![Screenshot 2023-08-06 at 12 42 31 PM](https://github.com/Cyfrin/foundry-full-course-f23/assets/44597990/6dbc66ff-2dec-431e-ad9f-aacbbdfe1dbc)

![Screenshot 2023-08-06 at 12 51 34 PM](https://github.com/Cyfrin/foundry-full-course-f23/assets/44597990/d7e08fb5-0852-4888-a286-425ba1bf412a)

![Screenshot 2023-08-06 at 12 51 51 PM](https://github.com/Cyfrin/foundry-full-course-f23/assets/44597990/c071f604-6316-4468-aef6-37093afa6a25)

So I would like to understand where I am going wrong .. As far as I understand Interfaces are standard way to make the design standard and they should include everything like IERC20 includes and then used further while implementing the token

It will be great if I can get some insights by @PatrickAlphaC Ser 🙏 as well along with other experts here 

Complete Code Ref : https://github.com/devrajsinghrawat/foundry-course

**Comment by shikhar229169** [comment, marked as answer]
When you are declared the withdraw and getVersion function inside of interface and when you inherit the interface inside the Test contract, that means you need to define the methods there, means you need to define the implementation in the Test contract, but if you are not implementing there then you need to mark Test contract as abstract. 

Abstract contract are ones in which we only have function definition but not implement them.
Suppose someone want to make their own implementation of Stack, so they will actually get the abstract contract of Stack in which there exists only function definition (but not implementation). And then they can inherit from the abstract stack contract in their main Stack contract, where they are required to implement all the functions for which only function definition is given in abstract contract (the functions which are not implemented).

The purpose of using abstract contract is that one knows what function are necessary to implement and they should implement all functions, because in abstract contract we will only provide function definition and then we need to provide their implementation in the main contract (in which we inherit the abstract contract). But if we don't implement even one function of abstract contract (for which only function definition was given in abstract contract) in our main contract then we need to make our main contract also as abstract.



Now coming back to the Test contract case, you were inheriting from IFundMe and for both of the functions their was no implementation thus it was required to be implemented in your Test contract otherwise it should be marked as abstract. And as abstract contract has only function definition but the functions are not implemented so abstract contract is never deployed.
And as Test contract was an abstract contract then it was not deployed, thus we got no matching test message.

**Reply by devrajsinghrawat** [reply]
Yes Agree what you said about the abstract contract So if I want to use Interface in this situation and then want only use event in the Test Contract then how should I expose that event from Interface to Test contract without making Test contract as Abstract contract.

**Reply by shikhar229169** [reply]
Then only add events in the interface contract and remove the functions which have no implementation defined (the one which you mentioned above).

**Reply by cromewar** [reply]
Amazing answer @shikhar229169
